plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,898 @@
+    termination_condition : `~plasmapy.simulation.particle_tracker.AbstractTerminationCondition`
+        A class responsible for determining when the simulation has reached
+        a suitable termination condition. The class is passed an instance
+        of `~plasmapy.simulation.particle_tracker.ParticleTracker`. See `~plasmapy.simulation.particle_tracker.AbstractTerminationCondition`
+        for more details.
+
+    save_routine : `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine`
+        An instance of `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine` which determines which
+        time steps of the simulation to save. See `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine` for more details.
+
+    dt : `~astropy.units.Quantity`, optional
+        An explicitly set time step in units convertible to seconds.
+        Setting this optional keyword overrules the adaptive time step
+        capability and forces the use of this time step throughout.
+
+    dt_range : tuple of shape (2,) of `~astropy.units.Quantity`, optional
+        If specified, the calculated adaptive time step will be clamped
+        between the first and second values.
+
+        field_weighting : str
+        String that selects the field weighting algorithm used to determine
+        what fields are felt by the particles. Options are:
+
+        * 'nearest neighbor': Particles are assigned the fields on
+            the grid vertex closest to them.
+        * 'volume averaged' : The fields experienced by a particle are a
+            volume-average of the eight grid points surrounding them.
+
+        The default is 'volume averaged'.
+
+    req_quantities : `list` of str, optional
+        A list of quantity keys required to be specified on the Grid object.
+        The base particle pushing simulation requires the quantities [E_x, E_y, E_z, B_x, B_y, B_z].
+        If any of these quantities are missing, a warning will be given and that
+        quantity will be assumed to be zero everywhere. This keyword is for specifying quantities in
+        addition to these six. The default is None.
+
+    verbose : bool, optional
+        If true, updates on the status of the program will be printed
+        into the standard output while running. The default is True.
+    """
+
+    def __init__(
+        self,
+        grids: Union[AbstractGrid, Iterable[AbstractGrid]],
+        termination_condition: AbstractTerminationCondition,
+        save_routine: Optional[AbstractSaveRoutine] = None,
+        dt=None,
+        dt_range=None,
+        field_weighting="volume averaged",
+        req_quantities=None,
+        verbose=True,
+    ):
+        # self.grid is the grid object
+        if isinstance(grids, AbstractGrid):
+            self.grids = [
+                grids,
+            ]
+        elif isinstance(grids, collections.abc.Iterable):
+            self.grids = grids
+        elif isinstance(grids, BasePlasma):
+            raise TypeError(
```

Comment:
Since we move the specification of the termination condition and the saving routine to the constructor this way of detecting old code no longer works and instead throws a `ValueError` because no termination condition is specified.. What do you recommend we do with this? @pheuer